### PR TITLE
Shaders

### PR DESCRIPTION
**Pull request by Octopus773**
## Proposed changes

Adding shaders component usable with ShaderComponentModel & ShaderComponentDrawable2D

## Information
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Non-breaking changes

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Review by Arthi-chaud — APPROVED** [review]

**Review by EternalRat — APPROVED** [review]

**Review by GitBluub — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,70 +18,73 @@ if (EMSCRIPTEN)
 endif ()
 
 set(SOURCES
-	sources/Models/GameState.hpp
```
Comment (AnonymusRaccoon) [inline CMakeLists.txt:21]: Keep tabs

File: lib/Ray/CMakeLists.txt [hunk 2]
```diff
@@ -97,7 +97,7 @@ set(SRC
     sources/Model/ModelAnimations.cpp
     sources/Vector/Vector2.cpp
     sources/Vector/Vector3.cpp
-)
+    sources/Shaders/Shaders.cpp sources/Shaders/Shaders.hpp)
```
Comment (AnonymusRaccoon) [inline lib/Ray/CMakeLists.txt:100]: Please keep the ')' on a new line, this make simpler merges.

File: lib/Ray/sources/Utils/Cache.hpp [hunk 3]
```diff
@@ -83,4 +85,42 @@ namespace RAY {
...
+		{};
+
+		std::shared_ptr<::Shader> fetch(const std::string &vertexFile, const std::string &fragmentFile)
+		{
+			const std::string index = vertexFile + fragmentFile;
+
+			if (vertexFile.empty() && fragmentFile.empty()) {
+				throw RAY::Exception::WrongInputError();
+			}
+			if (this->_cache.find(index) != this->_cache.end())
```
Comment (AnonymusRaccoon) [inline lib/Ray/sources/Utils/Cache.hpp:105]: You can use cache.contains(index)

File: sources/Component/Shaders/ShaderComponent.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+#include <Component/Component.hpp>
+#include <Entity/Entity.hpp>
+#include <Shaders/Shaders.hpp>
+
+namespace BBM
+{
+	class ShaderComponent : public WAL::Component
+	{
+	private:
+		//! @brief efefefefez
```
Comment (AnonymusRaccoon) [inline sources/Component/Shaders/ShaderComponent.hpp:17]: Nice documentation

File: sources/System/Renderer/RenderSystem.cpp [hunk 5]
```diff
@@ -31,16 +33,34 @@ namespace BBM
 		this->_window.clear();
 
 		this->_window.useCamera(this->_camera);
-		for (auto &[_, pos, drawable] : this->_wal.scene->view<PositionComponent, Drawable3DComponent>()) {
+		for (auto &[entity, pos, drawable] : this->_wal.scene->view<PositionComponent, Drawable3DComponent>()) {
+			auto *modelShader = entity.tryGetComponent<ShaderComponentModel>();
+
+			if (modelShader) {
+				auto &model = dynamic_cast<RAY::Drawables::Drawables3D::Model &>(*drawable.drawable);
```
Comment (AnonymusRaccoon) [inline sources/System/Renderer/RenderSystem.cpp:40]: Please store the dynamic casted on a variable, dynamic_cast is costly. Don't do it twice per frame.

**Review by AnonymusRaccoon — CHANGES_REQUESTED** [review]
Please refrain doing visual changes that does nothing else. It breaks git blame and make merges harder to do.

**Review by Arthi-chaud — APPROVED** [review]

**Pull request merged**